### PR TITLE
Flutter Web Bootstrapping Improvements

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
This makes several changes to flutter web app bootstrapping.
* The build now produces a `flutter_bootstrap.js` file.
  * By default, this file does the basic streamlined startup of a flutter app with the service worker settings and no user configuration.
  * The user can also put a `flutter_bootstrap.js` file in the `web` subdirectory in the project directory which can have whatever custom bootstrapping logic they'd like to write instead. This file is also templated, and can use any of the tokens  that can be used with the `index.html` (with the exception of `{{flutter_bootstrap_js}}`, see below).
* Introduced a few new templating tokens for `index.html`:
  * `{{flutter_js}}` => inlines the entirety of `flutter.js`
  * `{{flutter_service_worker_version}}` => replaced directly by the service worker version. This can be used instead of the script that sets the `serviceWorkerVersion` local variable that we used to have by default.
  * `{{flutter_bootstrap_js}}` => inlines the entirety of `flutter_bootstrap.js` (this token obviously doesn't apply to `flutter_bootstrap.js` itself).
* Changed `IndexHtml` to be called `WebTemplate` instead, since it is used for more than just the index.html now.
* We now emit warnings at build time for certain deprecated flows:
  * Warn on the old service worker version pattern (i.e.`(const|var) serviceWorkerVersion = null`) and recommends using `{{flutter_service_worker_version}}` token instead
  * Warn on use of `FlutterLoader.loadEntrypoint` and recommend using `FlutterLoader.load` instead
  * Warn on manual loading of `flutter_service_worker.js`.
* The default `index.html` on `flutter create` now uses an async script tag with `flutter_bootstrap.js`.

### CODE REVIEW

**Review by ditman — COMMENTED** [review]
Man, this looks GREAT. I'm updating one of my test apps to use this, see how it goes!

File: dev/a11y_assessments/web/index.html [hunk 1]
```diff
@@ -41,23 +41,8 @@
...
-        },
-        onEntrypointLoaded: function(engineInitializer) {
-          engineInitializer.initializeEngine().then(function(appRunner) {
-            appRunner.runApp();
-          });
-        }
-      });
-    });
-  </script>
+  <script src="flutter_bootstrap.js" asyn></script>
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply dev/a11y_assessments/web/index.html:46]: Whooops lol

**Review by ditman — COMMENTED** [review]
I have been testing this, and it's great. I wonder if the flutter_bootstrap.js file could be minified!

Also, somewhat unrelated to this PR (but IMO relevant for the feature); it'd be nice to have more feedback when the flutter bootstrap falls-back to JS, **why**.

Right now I have an application deployed that I'd expect to see in `--wasm`, but it's falling back to JS (in a supposedly supported browser), and I don't know why!

**Comment by ditman** [comment]
(Maybe a `_flutter.loader.log` variable that keeps the debugging information of the loader, without spitting it to the console?)

**Comment by eyebrowsoffire** [comment]
> Also, somewhat unrelated to this PR (but IMO relevant for the feature); it'd be nice to have more feedback when the flutter bootstrap falls-back to JS, **why**.
> 
> Right now I have an application deployed that I'd expect to see in `--wasm`, but it's falling back to JS (in a supposedly supported browser), and I don't know why!

Absolutely. Kevin also gave me that feedback (well, specifically that we should spit out more information the `crossOriginIsolated` case). I do plan on doing that, but it would be a PR in the engine, not the framework, since it would be code inside `flutter.js` itself that would emit this info.

**Comment by eyebrowsoffire** [comment]
The issue Kevin filed about more info about the JS fallback is here: https://github.com/flutter/flutter/issues/142822

**Comment by ditman** [comment]
Ah, another bit of feedback relevant to this issue that I think would be easy to implement:

I'd like to be able to set `{{flutter_js}}` and `{{flutter_build_config}}` in any order. I think the `flutter_build_config` just needs a small prefix where it defines `_flutter` if it isn't defined yet, similar to here:

* https://github.com/flutter/engine/blob/main/lib/web_ui/flutter_js/src/flutter.js#L7-L11

(We need a change to flutter.js for that though, grrr...)

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 2]
```diff
@@ -328,18 +329,32 @@ class Dart2WasmTarget extends Dart2WebTarget {
...
+
+  static String generateBuildConfigString(List<Dart2WebTarget> compileTargets) {
+    final List<Map<String, Object?>> buildDescriptions = compileTargets.map(
+      (Dart2WebTarget target) => target.buildConfig
+    ).toList();
+    final Map<String, Object?> buildConfig = <String, Object?>{
+      'engineRevision': globals.flutterVersion.engineRevision,
+      'builds': buildDescriptions,
+    };
+    return 'if (!_flutter) _flutter = {}; _flutter.buildConfig = ${jsonEncode(buildConfig)};';
```
Comment (ditman) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart:353]: This syntax caused an issue when I was doing a similar change in the engine. Let me give this a shot with a custom bootstrap file (rendering the config before flutter.js)

File: dev/integration_tests/web_e2e_tests/web/flutter_bootstrap.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{{flutter_js}}
+{{flutter_build_config}}
```
Comment (ditman) [inline dev/integration_tests/web_e2e_tests/web/flutter_bootstrap.js:2]:
```suggestion
{{flutter_build_config}}
{{flutter_js}}
```

File: packages/flutter_tools/lib/src/build_system/targets/web.dart (hunk 2 above)
Comment (ditman) [reply packages/flutter_tools/lib/src/build_system/targets/web.dart:353]: Yep:

```
Uncaught ReferenceError: _flutter is not defined
    at (index):59:5
```

Comment (ditman) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart:353]:
```suggestion
    return '''
if (!window._flutter) {
  window._flutter = {};
}
_flutter.buildConfig = ${jsonEncode(buildConfig)};
''';
```

(Not sure I'm doing multi-line strings right :P)

File: packages/flutter_tools/lib/src/isolated/devfs_web.dart [hunk 4]
```diff
@@ -536,19 +539,47 @@ class WebAssetServer implements AssetReader {
         },
       ],
     };
-    final String buildConfigString = '_flutter.buildConfig = ${jsonEncode(buildConfig)};';
+    return 'if (!_flutter) _flutter = {}; _flutter.buildConfig = ${jsonEncode(buildConfig)};';
```
Comment (ditman) [inline packages/flutter_tools/lib/src/isolated/devfs_web.dart:542]:
```suggestion
    return '''
if (!window._flutter) {
  window._flutter = {};
}
_flutter.buildConfig = ${jsonEncode(buildConfig)};
''';
```

File: packages/flutter_tools/lib/src/web/bootstrap.dart [hunk 5]
```diff
@@ -472,3 +472,16 @@ String generateTestBootstrapFileContents(
 })();
 ''';
 }
+
+String generateDefaultFlutterBootstrapScript() {
+  return '''
+{{flutter_js}}
+{{flutter_build_config}}
```
Comment (ditman) [inline packages/flutter_tools/lib/src/web/bootstrap.dart:479]:
```suggestion
{{flutter_build_config}}
{{flutter_js}}
```

Comment (eyebrowsoffire) [reply packages/flutter_tools/lib/src/web/bootstrap.dart:479]: Let's keep the default bootstrap as-is, we don't need to change the order.

**Review by ditman — APPROVED** [review]
Let's go!

**Pull request merged**